Add support for streaming S3 file uploads

jaxwilko opened the pull request:
This PR adds support for streaming S3 uploads from `MediaManager` & `FileUpload` (todo: `RichEditor`). These changes have been implemented using `Events` that will need to be added to the winter core. I will link the core PR here when it's up.

jaxwilko committed:
Added controller for generating signed urls

jaxwilko committed:
Added js for overriding default upload handling

jaxwilko committed:
Added config options

jaxwilko committed:
Added git ignore

jaxwilko committed:
Added event hooks to support streaming file uploads to s3

jaxwilko commented on package.json:
```diff
@@ -0,0 +1,38 @@
+{
```
No idea how this got here, will remove

jaxwilko committed:
Fixed package.json

jaxwilko committed:
Added froala stream support override

jaxwilko committed:
Added richeditor support

jaxwilko committed:
Added richeditor support and simplified overrides

jaxwilko committed:
Switched controller for behaviour

jaxwilko committed:
Added config options

jaxwilko committed:
Cleaned up js and implemented fixes

LukeTowers committed:
Initial code review

LukeTowers commented on Plugin.php:
```diff
@@ -97,4 +136,143 @@ public function boot()
+                 * - path: The path to put the uploaded file (relative to the media folder and only takes effect if $widget->uploadPath is not set)
+                 */
+                $uploadedPath = 'tmp/' . Request::get('uuid');
+                $originalName = Request::get('name');
```
@jaxwilko why are you accessing these input vars through `Request::get()` instead of `Request::input()` or even `Request::post()`?

jaxwilko replied:
Think it's a copy paste from somewhere else that was using that style

LukeTowers committed:
Add temporary comments

LukeTowers committed:
Use the provided file model instance

LukeTowers committed:
Add link to AWS SDK issue in progress

LukeTowers committed:
Minor cleanup

LukeTowers committed:
Simplify processing of UploadableWidget uploads

Also move enabling of the s3 streamable uploads functionality to the individual disk configurations.

LukeTowers committed:
Code cleanup

LukeTowers committed:
Fix detection of streamable uploads support

jaxwilko committed:
Added fix for form property propagation to froala

jaxwilko committed:
Compiled stream plugin

LukeTowers committed:
Refactor SignedStorageUrlBehaviour

Inspiration taken from https://github.com/laravel/vapor-core/blob/2.0/src/Http/Controllers/SignedStorageUrlController.php

LukeTowers committed:
Use the ContentType set by the client

Switch to using a language string for the file size exceeded error message and removed the now unnecessary config file.

Also explicitly set the uploaded file's file name and content type for FileUploads

LukeTowers committed:
Add language string for failed upload

LukeTowers committed:
Add author block

LukeTowers committed:
Add documentation

LukeTowers committed:
Updated JS event names

LukeTowers committed:
Add disclaimer
